### PR TITLE
Enable log collection for E2E tests

### LINKED ISSUE #734: dump logs from CAPI and CAPV components into the artifacts directory
/kind feature

**Describe the solution you'd like**
Currently our e2e test do not dump logs from CAPI and CAPV, this makes hard to debug things while running on CI. 
We should invoke `kubectl  cluster-info dump` and add the content to the artifacts directory before garbage collecting the kind cluster 

**Anything else you would like to add:**

/assign @yastij 
/priority important-soon

**Environment:**

- Cluster-api-provider-vsphere version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by srm09**
**What this PR does / why we need it**:
This patch enables log collection for E2E tests

**Which issue(s) this PR fixes**:
Fixes #734

**Special notes for your reviewer**:
n/a

**Release note**:
```release-note
Enables log collection for E2E tests
```

### CODE REVIEW

**Comment by srm09** [comment]
/cc @gab-satchi @yastij PTAL. 

Need some information around the useful logs to be captured.

File: test/e2e/log_collector.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+	"os"
+	"path/filepath"
+	"strings"
+
+	"golang.org/x/crypto/ssh"
+
+	"github.com/pkg/errors"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	//kinderrors "sigs.k8s.io/kind/pkg/errors"
```
Comment (yastij) [inline test/e2e/log_collector.go:33]: should we remove this ?

File: test/e2e/log_collector.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+func createOutputFile(path string) (*os.File, error) {
+	if err := os.MkdirAll(filepath.Dir(path), os.ModePerm); err != nil {
+		return nil, err
+	}
+	return os.Create(path)
+}
+
+func executeRemoteCommand(f io.StringWriter, hostIpAddr, command string, args ...string) error {
```
Comment (yastij) [inline test/e2e/log_collector.go:74]: nit: `hostIpAddr` should be `hostIPAddr`

File: test/e2e/log_collector.go (hunk 1 above)
Comment (srm09) [reply test/e2e/log_collector.go]: Using this now.

**Comment by srm09** [comment]
/retitle Enable log collection for E2E tests

File: test/e2e/log_collector.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
+/*
+Copyright 2020 The Kubernetes Authors.
```
Comment (gab-satchi) [inline test/e2e/log_collector.go:2]: nit: 2021

**Review by gab-satchi — COMMENTED** [review]
changes look good. We should bring in a secret `VSPHERE_SSH_PRIVATE_KEY` from test-infra and validate.

**Comment by fabriziopandini** [comment]
SGTM for me
Great to see the E2E test framework being used!

File: test/e2e/log_collector.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+	"golang.org/x/crypto/ssh"
+
+	"github.com/pkg/errors"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	kinderrors "sigs.k8s.io/kind/pkg/errors"
+)
+
+const (
+	VSpherePrivateKey = "VSPHERE_SSH_PRIVATE_KEY"
```
Comment (yastij) [inline test/e2e/log_collector.go:37]: from looking at the e2e artifacts directory, it seems like we're collecting stuff (e.g. cloud-init-output.log etc..). which ssh key are we using ?

**Review by yastij — COMMENTED** [review]
/approve

lgtm, just one question

File: test/e2e/log_collector.go (hunk 4 above)
Comment (srm09) [reply test/e2e/log_collector.go:37]: The idea is to add a new ssh key to the prow job which is used to run the tests.

Comment (yastij) [reply test/e2e/log_collector.go:37]: @srm09 - how are we currently getting [this](https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-vsphere/1110/pull-cluster-api-provider-vsphere-e2e/1359984816526725120/artifacts/clusters/quick-start-y2vn14/machines/quick-start-y2vn14-rt598/) without an ssh key ?

Comment (gab-satchi) [reply test/e2e/log_collector.go:37]: They're all empty files. The changes create the file before attempting to execute the command

**Comment by srm09** [comment]
Secrets have been added: https://github.com/kubernetes/test-infra/issues/21099

**Comment by gab-satchi** [comment]
Thanks Sagar. I see the private key brought into the prow job but it looks like your changes require an env var set?

**Pull request merged**